File: src/python/ch01/1.3/l2/extgcd_inv.py
from typing import Tuple

def extended_gcd(a: int, b: int) -> Tuple[int, int, int]:
    """
    Return (g, x, y) such that a*x + b*y = g = gcd(a, b).
    Iterative version; x, y are Bézout coefficients.
    """
    old_r, r = a, b
    old_s, s = 1, 0
    old_t, t = 0, 1
    while r:
        q = old_r // r
        old_r, r = r, old_r - q * r
        old_s, s = s, old_s - q * s
        old_t, t = t, old_t - q * t
    # old_r is gcd, and a*old_s + b*old_t = old_r
    return old_r, old_s, old_t

def invmod(a: int, m: int) -> int:
    """
    Return modular inverse x with (a*x) % m == 1, raising ValueError if not coprime.
    Result is in range [0, m-1].
    """
    if m <= 0:
        raise ValueError("modulus must be positive")
    g, x, _ = extended_gcd(a, m)
    if g != 1 and g != -1:
        raise ValueError("inverse does not exist; gcd(a, m) != 1")
    return x % m

File: src/go/ch01/1.3/l2/extgcd_inv.go
package main

import "errors"

// ExtendedGCD returns (g, x, y) such that a*x + b*y = g = gcd(a,b).
func ExtendedGCD(a, b int64) (g, x, y int64) {
	oldR, r := a, b
	oldS, s := int64(1), int64(0)
	oldT, t := int64(0), int64(1)
	for r != 0 {
		q := oldR / r
		oldR, r = r, oldR-q*r
		oldS, s = s, oldS-q*s
		oldT, t = t, oldT-q*t
	}
	return oldR, oldS, oldT
}

// InvMod returns x in [0,m-1] with (a*x) % m == 1; error if gcd(a,m) != 1.
func InvMod(a, m int64) (int64, error) {
	if m <= 0 {
		return 0, errors.New("modulus must be positive")
	}
	g, x, _ := ExtendedGCD(a, m)
	if g != 1 && g != -1 {
		return 0, errors.New("inverse does not exist; gcd(a,m) != 1")
	}
	// normalize
	x %= m
	if x < 0 {
		x += m
	}
	return x, nil
}